File: backend/app/schemas/revisions.py
from __future__ import annotations

from datetime import datetime
from typing import Any

from pydantic import BaseModel, ConfigDict, Field

from app.models.config_revision import DeploymentStatus, ValidationStatus


class RevisionSummary(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    service_instance_id: str
    revision_number: int
    validation_status: ValidationStatus
    deployment_status: DeploymentStatus
    created_by: str
    created_at: datetime
    deployed_at: datetime | None = None


class RevisionRead(RevisionSummary):
    configuration: dict[str, Any]
    rendered_configuration: str
    validation_output: str
    diff_from_previous: str | None = Field(
        default=None,
        description="Unified diff of rendered config against previous revision",
    )

File: frontend/src/types/revisions.ts
import { z } from "zod";

export const RevisionSummarySchema = z.object({
  id: z.string(),
  service_instance_id: z.string(),
  revision_number: z.number().int(),
  validation_status: z.enum(["valid", "invalid", "unknown"]),
  deployment_status: z.enum(["pending", "deployed", "failed", "superseded"]),
  created_by: z.string(),
  created_at: z.string(),
  deployed_at: z.string().nullable().optional(),
});

export const RevisionReadSchema = RevisionSummarySchema.extend({
  configuration: z.record(z.string(), z.unknown()),
  rendered_configuration: z.string(),
  validation_output: z.string(),
  diff_from_previous: z.string().nullable().optional(),
});

export type RevisionSummary = z.infer<typeof RevisionSummarySchema>;
export type RevisionRead = z.infer<typeof RevisionReadSchema>;
